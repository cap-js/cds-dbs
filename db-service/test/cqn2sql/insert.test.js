'use strict'
const cds = require('@sap/cds/lib')
const cqn2sql = require('../../lib/cqn2sql')

beforeAll(async () => {
  cds.model = await cds.load(__dirname + '/testModel').then(cds.compile.for.nodejs)
})

describe('insert', () => {
  describe('insert only', () => {
    // Values are missing
    test('test with insert values into columns', () => {
      const cqnInsert = {
        INSERT: {
          into: { ref: ['Foo'] },
          columns: ['ID', 'b', 'x'],
          values: [1, "'asd'", 2],
        },
      }

      const { sql, entries } = cqn2sql(cqnInsert)
      expect({ sql, entries }).toMatchSnapshot()
    })

    test('test with insert rows into columns', () => {
      const cqnInsert = {
        INSERT: {
          into: { ref: ['Foo'] },
          columns: ['ID', 'b', 'a'],
          rows: [
            [1, "'asd'", 2],
            [9, "mmm'", 77],
          ],
        },
      }
      const { sql, entries } = cqn2sql(cqnInsert)
      expect({ sql, entries }).toMatchSnapshot()
    })

    // no filtering in INSERT
    xtest('test filter in insert rows into columns with not existing column', () => {
      const cqnInsert = {
        INSERT: {
          into: { ref: ['Foo2'] },
          columns: ['ID', 'not_existing', 'something'],
          rows: [
            [1, "'asd'", 2],
            [9, "mmm'", 77],
          ],
        },
      }
      const { sql, entries } = cqn2sql(cqnInsert)
      expect({ sql, entries }).toMatchSnapshot()
    })

    test('test with insert entries', () => {
      const cqnInsert = {
        INSERT: {
          into: 'Foo2',
          entries: [
            { ID: 1, name: null, a: 2 },
            { ID: null, name: "'asd'", a: 6 },
          ],
        },
      }

      const { sql, entries } = cqn2sql(cqnInsert)
      expect({ sql, entries }).toMatchSnapshot()
    })

    test('test with insert with alias', () => {
      const cqnInsert = {
        INSERT: {
          into: { ref: ['Foo2'], as: 'Fooooo2' },
          entries: [
            { ID: 1, name: null, a: 2 },
            { ID: null, name: "'asd'", a: 6 },
          ],
        },
      }

      const { sql, entries } = cqn2sql(cqnInsert)
      expect({ sql, entries }).toMatchSnapshot()
    })
  })

  describe('insert into ... select from ...', () => {
    // sql is generated correctly, but not valid since number of columns is different in both tables
    test('no columns', () => {
      const cqnInsert = {
        INSERT: {
          into: 'Foo',
          as: { SELECT: { from: { ref: ['Foo2'] } } },
        },
      }

      const { sql } = cqn2sql(cqnInsert)
      expect(sql).toEqual('INSERT INTO Foo (ID,a,b,c,x) SELECT Foo2.ID,Foo2.name,Foo2.a FROM Foo2 as Foo2')
    })

    test('with columns', () => {
      const cqnInsert = {
        INSERT: {
          into: 'Foo',
          columns: ['ID'],
          as: { SELECT: { from: { ref: ['Foo2'] }, columns: [{ ref: ['ID'] }] } },
        },
      }

      const { sql } = cqn2sql(cqnInsert)
      expect(sql).toEqual('INSERT INTO Foo (ID) SELECT Foo2.ID FROM Foo2 as Foo2')
    })
  })
})
