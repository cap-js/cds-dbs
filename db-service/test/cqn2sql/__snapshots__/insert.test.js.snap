// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert insert only test filter in insert rows into columns with not existing column 1`] = `
{
  "entries": [
    [
      "[[1,"'asd'",2],[9,"mmm'",77]]",
    ],
  ],
  "sql": "INSERT INTO Foo2 (ID,not_existing,something) SELECT value->>'$[0]',value->>'$[1]',value->>'$[2]' FROM json_each(?)",
}
`;

exports[`insert insert only test with insert entries 1`] = `
{
  "entries": [
    [
      "[{"ID":1,"name":null,"a":2},{"ID":null,"name":"'asd'","a":6}]",
    ],
  ],
  "sql": "INSERT INTO Foo2 (ID,name,a) SELECT value->>'$."ID"',value->>'$."name"',value->>'$."a"' FROM json_each(?)",
}
`;

exports[`insert insert only test with insert rows into columns 1`] = `
{
  "entries": [
    [
      "[[1,"'asd'",2],[9,"mmm'",77]]",
    ],
  ],
  "sql": "INSERT INTO Foo (ID,b,a) SELECT value->>'$[0]',value->>'$[1]',value->>'$[2]' FROM json_each(?)",
}
`;

exports[`insert insert only test with insert values into columns 1`] = `
{
  "entries": [
    [
      "[[1,"'asd'",2]]",
    ],
  ],
  "sql": "INSERT INTO Foo (ID,b,x) SELECT value->>'$[0]',value->>'$[1]',value->>'$[2]' FROM json_each(?)",
}
`;

exports[`insert insert only test with insert with alias 1`] = `
{
  "entries": [
    [
      "[{"ID":1,"name":null,"a":2},{"ID":null,"name":"'asd'","a":6}]",
    ],
  ],
  "sql": "INSERT INTO Foo2 as Fooooo2 (ID,name,a) SELECT value->>'$."ID"',value->>'$."name"',value->>'$."a"' FROM json_each(?)",
}
`;
